### ISSUE TITLE
Compiling native image got error caused by automatic initialization of safe-class

### ISSUE THREAD

**Issue by mayeths** [oracle/graal#3587]
### Describe the issue
**TL; DR: Can we disable safe-class initialization by options or something else?**

Explain:

We are trying to make a maven plugin that generate AOT information to help user compile their project into native image. Information is about these projects' [dubbo](https://github.com/apache/dubbo) dependency, a RPC framework, because dubbo use reflection and dynamic code generating (like with javassist) to create [adapter](https://en.wikipedia.org/wiki/Adapter_pattern) classes for extensions at run-time, even for many dubbo's built-in components. In graalvm world, these adapter classes should be created at AOT phase, since graalvm doesn't support generating classes dynamically.

Now generating adapter works well, but the problem occurred: compiling project with dubbo into native-image got error like this [nativedemo](https://github.com/mayeths/nativedemo). For this tiny demo, the problem is `org.eclipse.collections.impl.set.immutable.ImmutableEmptySet` got build-time initialization and its dependency `org.eclipse.collections.api.factory.Sets` got run-time initialization. The phenomenon is associated with safe-class checking imo (see **More details** bellow).

**So I wonder how can we disable [Automatic Initialization of Safe Classes](https://www.graalvm.org/reference-manual/native-image/ClassInitialization/#automatic-initialization-of-safe-classes) (or for some packages, so we can force packages associated with dubbo to initiate at run-time)?** Dubbo is really a big project, and user may meet some build/run-time initialization conflicts caused by dubbo's dependencies. We can `--trace-class-initialization` everytime an exception was thrown, but it's somewhat ugly for our maven plugin (no out-of-the-box).

### Steps to reproduce the issue
The code was posted to nativedemo.
```
git clone https://github.com/mayeths/nativedemo
cd nativedemo
# Follow the 'Prerequisite', 'Normal Build & Run' and 'Native Image Build' sections in README to reproduce
```

### Environment infomation
- Native Image
```
$ native-image --version
GraalVM 21.1.0 Java 11 CE (Java Version 11.0.11+8-jvmci-21.1-b05)
```
- Maven
```
$ mvn -v
Apache Maven 3.6.0
Maven home: /usr/share/maven
Java version: 11.0.11, vendor: GraalVM Community, runtime: /home/mayeths/module/graalvm-ce-java11-21.1.0
Default locale: en, platform encoding: UTF-8
OS name: "linux", version: "4.4.0-18362-microsoft", arch: "amd64", family: "unix"
```

### More details

In nativedemo:

- You can see `org.eclipse.collections.impl.set.immutable.ImmutableEmptySet` was marked as safe-class in `consumer/target/reports/safe_classes_DATE_TIME.txt` and appeared in `consumer/target/reports/build_time_classes_DATE_TIME.txt`.

- But its dependency `org.eclipse.collections.api.factory.Sets` was appeared in `consumer/target/reports/run_time_classes_DATE_TIME.txt` (filled red color in `consumer/target/reports/initializer_dependencies_DATE_TIME.dot`).

- Adding `--initialize-at-build-time=org.eclipse.collections.api.factory.Sets` to `pom.xml` will make it correct. **But running native-image executable will failed.**

**Comment by oubidar-Abderrahim** [comment]
Hi, thank you for reporting this, we will take a look into it shortly

**Comment by christianwimmer** [comment]
Disabling initialization of safe classes is certainly not a solution. Because, by definition, there must not be an observable side effect of such an initialization. We need to look into your example, maybe there is a bug and a class gets automatically initialized even though it is not safe.

Can you please set a breakpoint in the class initializer of `org.eclipse.collections.api.factory.Sets`, and post the stack trace showing why it gets executed at image build time? Or use `--trace-class-initialization` to get that stack trace?

**Comment by mayeths** [comment]
> Disabling initialization of safe classes is certainly not a solution. Because, by definition, there must not be an observable side effect of such an initialization. We need to look into your example, maybe there is a bug and a class gets automatically initialized even though it is not safe.
> 
> Can you please set a breakpoint in the class initializer of `org.eclipse.collections.api.factory.Sets`, and post the stack trace showing why it gets executed at image build time? Or use `--trace-class-initialization` to get that stack trace?

Yes, I added `--trace-class-initialization` to args in pom.xml, so the section [_Native Image Build_](https://github.com/mayeths/nativedemo#native-image-build) descripbed the stack trace:
```
mvn clean install -DskipTests -Pnative

# ...
# [consumer:28982]      compile:  37,777.86 ms,  5.88 GB
# [consumer:28982]        image:   4,029.36 ms,  5.87 GB
# [consumer:28982]        write:     827.04 ms,  5.87 GB
# Error: Classes that should be initialized at run time got initialized during image building:
#  org.eclipse.collections.api.factory.Sets was unintentionally initialized at build time.
org.eclipse.collections.impl.set.immutable.ImmutableEmptySet
caused initialization of this class with the following trace: 
#         at org.eclipse.collections.api.factory.Sets.<clinit>(Sets.java:47)
#         at org.eclipse.collections.impl.set.mutable.UnifiedSet.toImmutable(UnifiedSet.java:996)
#         at org.eclipse.collections.impl.partition.set.PartitionImmutableSetImpl.<init>(PartitionImmutableSetImpl.java:24)
#         at org.eclipse.collections.impl.partition.set.PartitionUnifiedSet.toImmutable(PartitionUnifiedSet.java:38)
#         at org.eclipse.collections.impl.set.immutable.ImmutableEmptySet.<clinit>(ImmutableEmptySet.java:53)
# ...
```

**Comment by christianwimmer** [comment]
I can confirm this is a bug in the automatic initialization of classes, I'm working on a fix.

**Comment by mayeths** [comment]
> I can confirm this is a bug in the automatic initialization of classes, I'm working on a fix.

Thanks, I will check it out when this patch is available.
